\documentclass[aspectratio=169,11pt]{beamer}

\usetheme{Singapore}
\usepackage[utf8]{inputenc}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{hyperref}

% Setup the bibliography
\usepackage[style=authortitle,backend=bibtex]{biblatex}
\addbibresource{bibliography.bib}
\setbeamertemplate{bibliography item}[text]
\setbeamerfont{footnote}{size=\tiny}

% Allow footnotes with no number
\newcommand\blfootnote[1]{%
  \begingroup
  \renewcommand\thefootnote{}\footnote{#1}%
  \addtocounter{footnote}{-1}%
  \endgroup
}

% Allow section title slides
\AtBeginSection[]{
  \begin{frame}
  \vfill
  \centering
  \begin{beamercolorbox}[sep=8pt,center,shadow=true,rounded=true]{title}
    \usebeamerfont{title}\insertsectionhead\par%
  \end{beamercolorbox}
  \vfill
  \end{frame}
}

\author{Dr Stephen Pederson}
\title{Lecture 2: Early Transcriptomic Strategies}
\subtitle{BIOINF3005/7160: Transcriptomics Applications}
%\setbeamercovered{transparent} 
\setbeamertemplate{navigation symbols}{} 
\logo{
	\includegraphics[scale=0.3]{figures/UoA_logo_col_vert.png} 
} 
\institute{Bioinformatics Hub, \\The University of Adelaide} 
\date{March 16th, 2020} 
\subject{BIOINF3005/7160: Transcriptomics Applications} 


\begin{document}

\begin{frame}
\titlepage
\end{frame}

\begin{frame}
\footnotesize
\tableofcontents
\end{frame}

\section{Overview}

\linespread{1.4}

\begin{frame}{The Motivation}

	\begin{itemize}
		\item The transcriptome is a highly dynamic set of molecules
		\item Small changes can potentially have significant ramifications
		\begin{itemize}
			\item e.g. a ``Master Regulator" can determine cellular fate
		\end{itemize}
		\item RNA molecules are small
		\begin{itemize}
			\item How do we find what's in our sample?
			\item How do we quantify RNA?
			\item And how do we compare one or more groups?
		\end{itemize}
	\end{itemize}

\end{frame}

\begin{frame}{Technological Developments}

	\begin{itemize}
		\item Technological developments are constant
		\item Technologies are often transient
		\item Key technologies are:
		\begin{enumerate}
			\item Real Time Polymerase Chain Reaction (RT-PCR)
			\item Expressed Sequence Tags (EST)
			\item Serial/Cap Analysis of Gene Expression (SAGE/CAGE)
			\item Microarray technologies
			\item Sequencing technologies
		\end{enumerate}
		\item Analytic methodologies \textit{often lag technologies}
	\end{itemize}

\end{frame}

\begin{frame}{A Simplified History}

	\begin{center}
	\includegraphics[scale=0.5]{figures/timeTrends.png}		
	~\\
	\textcolor[rgb]{0.1,0.4,0.7}{EST (blue)};
	\textcolor[rgb]{0.85,0.65,0.15}{SAGE / CAGE (yellow)};
	\textcolor[rgb]{0.7,0,0.2}{Microarrays (red)};  
	RNA Seq (black) \footfullcite{2017transcriptomictech}
	\end{center}
 

\end{frame}

\section{Measuring Single Genes}

%\subsection{Northern Blot}

\begin{frame}{The Northern Blot}

	\begin{itemize}
		\item One of the earliest strategies\footfullcite{1977NorthernBlot}
		\item Developed as an extension of the Southern Blot\footfullcite{pmid1195397} (DNA)
		\item Gel Electrophoresis-based strategy
		\begin{itemize}
			\item Based on size differentiation and probe sequences
		\end{itemize}
	\end{itemize}

\end{frame}

\begin{frame}{The Northern Blot}

	\begin{itemize}
		\item RNA is extracted then denatured
		\item RNA is size separated using Gel Electrophoresis
		\item RNA is transferred to a ``blotting membrane"
		\item Treat the membrane with a labelled probe
		\begin{itemize}
			\item Probes are complementary to the ``target sequence"
			\item Probes are labelled with fluorescent dye or radioactive atoms
		\end{itemize}
	\end{itemize}

\end{frame}

\begin{frame}{The Northern Blot}

	\begin{center}
		\includegraphics[scale=0.3]{figures/northern.png} 
	\end{center}

\blfootnote{Taken from \cite{pmid25648206}}	

\end{frame}

\begin{frame}{The Northern Blot}
	\begin{itemize}
		\item Prominent usage \textit{before} genomes were sequenced
		\item Can possibly detect different isoforms
		\item Crude quantitation using Densitometric Analysis
		\begin{itemize}
			\item What limitations might this have?
			% \item Not accurate for low abundance
			% \item Very time intensive
			% \item Requires large amounts of starting material
		\end{itemize}
	\end{itemize}
\end{frame}

%\subsection{RT-PCR}

\begin{frame}{RT-qPCR}

	\begin{itemize}
		\item Reverse Transcriptase quantitative PCR
		\begin{itemize}
			\item Sometimes called: qPCR, RT-PCR
		\end{itemize}
		\item Often considered to be the ``gold standard" for quantitation
		\item Targets a \textit{specific transcribed region} via \textit{specific primers}
		\begin{itemize}
			\item Primers must be individually designed
			\item Primers often span exon-exon junctions
		\end{itemize}
	\end{itemize}

\end{frame}

\begin{frame}{RT-qPCR}

	\begin{enumerate}
		\item \textit{Reverse Transcriptase} converts RNA to cDNA
		\begin{itemize}
			\item Primers are required: Can target poly-A or random
		\end{itemize}
		\item Sequence-specific primers amplify the target fragment in cycles
		\begin{itemize}
			\item Fluorescent dye is commonly incorporated during amplification
		\end{itemize}
		\item Abundance of target will grow exponentially ($\times 2$) for each amplification cycle
		\item The cycle where abundance reaches the ``limit of detection" is estimated ($C_T$)
	\end{enumerate}

\end{frame}

\begin{frame}{RT-qPCR}

	\begin{center}
	\includegraphics[scale=0.3]{figures/RTqPCR1.png} 
	\end{center}
	
	\blfootnote{Modified from \cite{By Lokeshthimmana - Own work, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=76313637}}	

\end{frame}

\begin{frame}{RT-qPCR}

	\begin{center}
	\includegraphics[scale=0.3]{figures/RTqPCR2.png} 
	\end{center}
	
	\blfootnote{Modified from \cite{By Lokeshthimmana - Own work, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=76313637}}	

\end{frame}

\begin{frame}{RT-qPCR}

	\begin{center}
		\includegraphics[scale=0.5]{figures/qPCR.png} 
		~\\
		This is a 10-fold dilution series\footfullcite{2013qPCR}
	\end{center}

\end{frame}

\begin{frame}{RT-qPCR}

	\begin{itemize}
		\item Can be used with a standard curve and dilution series to estimate absolute quantity of an RNA \textit{within a sample}
		\item Can be used to compare \textit{across samples for relative abundance}
	\end{itemize}
	\pause
	\begin{center}
		~\\
		\textit{What may be a fundamental issue when comparing across samples?}
	\end{center}


\end{frame}

\begin{frame}{Normalisation}

	\begin{itemize}
		\item There may be pipetting and other technical differences between samples
		\begin{itemize}
			\item These are \textbf{non-biological} in origin
		\end{itemize}
		\item To correct for these we can \textbf{normalise} our data
		\item In RT-qPCR this is often done using ``housekeeper" genes
		\begin{itemize}
			\item We choose genes which \textit{should not} change between samples/groups
			\item These are commonly structural genes such as \textit{ACTN$\beta$} or \textit{GAPDH}
		\end{itemize}
	\end{itemize}

\end{frame}

\begin{frame}{Estimating Change In Expression}

	\begin{itemize}
		\item Relative abundances are often referred to as fold-change (FC)
		\begin{itemize}
			\item Down regulation is squeezed between 0 and 1
			\item Up regulation ranges from 1 to $\infty$
		\end{itemize}
		
		\item We often use $\log_2$ fold-change to get a better scale, e.g. 
		\begin{itemize}
			\item A 2-fold increase in abundance: $\log_2 2^1=1$
			\item A 2-fold decrease in abundance: $\log_2 \frac{1}{2}= \log_2 2^{-1} = -1$
			\item No change in abundance $\log_2 1 = \log_2 2^0 = 0$
		\end{itemize}		 
		\item This is often abbreviated as \textit{logFC}
	\end{itemize}

\end{frame}

\begin{frame}{Estimating Change In Expression}

	\begin{itemize}
		\item For \textit{RT-qPCR} the estimate of \textit{logFC} is known as $\Delta\Delta C_T$
		\item To calculate this, we calculate \textbf{two} changes in $C_T$
		\begin{enumerate}
			\item $\Delta C_T$ relative to the housekeeper(s)
			\item $\Delta \Delta C_T$ across samples for our gene/fragment of interest 
		\end{enumerate}
		\item The first step corrects for technical errors
		\item The second step estimates our true change in abundance
	\end{itemize}

\end{frame}

\begin{frame}{Estimating Change In Expression}

	Within each sample

	\begin{align*}
	\Delta C_T = C_{t[\text{gene}]} - C_{t[\text{HK}]}
	\end{align*}
	
	Across samples/groups
	
	\begin{align*}
	\Delta \Delta C_T = - (\Delta C_{T[\text{group1}]} - \Delta C_{T[\text{group2}]})
	\end{align*}
	
	This formulation assumes \textit{equal amplification efficiency} for all primers/genes (i.e. \textit{Efficiency} $= 2$)

\end{frame}

\begin{frame}{Estimating Change In Expression}

	\begin{itemize}
		\item Housekeeper genes must be matched to the ``gene of interest" \textbf{within each sample} and \textbf{within each qPCR reaction}
		\item Choosing $>1$ housekeeper gene is advised
		\item Measurements are often taken in triplicate/quadruplicate for each sample (reactions sometimes fail)
	\end{itemize}
	
	~\\
	\textit{Both Northern blots and RT-qPCR use targeted primers, but in very different ways}

\end{frame}

\section{Measuring Multiple Genes}

%\subsection{EST}

\begin{frame}{Expressed Sequence Tags}

	\begin{itemize}
		\item The first attempt at capturing the larger transcriptome was via Expressed Sequence Tags\footfullcite{1991VenterEST} (ESTs) in 1991
		\begin{itemize}
			\item Sequenced 609 mRNA human brain mRNA sequences
			\item ESTs were generated by reverse transcribing poly-A selected mRNA, amplified using random primers
			\item Used ESTs $\sim 100-800$nt
			\item Obtained actual sequences using Sanger Sequencing
		\end{itemize}
		\item $>$10 years before the Human Genome Project completed
		\item Just \textbf{discovering} genes was a huge priority
	\end{itemize}

\end{frame}

\begin{frame}{Expressed Sequence Tags}

	\begin{center}
	\includegraphics[scale=0.2]{figures/sanger.png} 
	\end{center}
	
	\blfootnote{Source: \url{https://www.khanacademy.org/science/high-school-biology/hs-molecular-genetics/hs-biotechnology/a/dna-sequencing}}

\end{frame}

\begin{frame}{Serial Analysis of Gene Expression}

\textit{Serial Analysis of Gene Expression}\footfullcite{pmid7570003} (SAGE) was the first attempt to quantify expression on a larger scale

	\begin{enumerate}
		\item Conversion of mRNA to ds-cDNA using biotinylated primers (often poly-T)
		\item cDNA is bound to beads using biotin and cleaved
		\item 11-mer ``tags" were produced after cleavage and concatenated
		\item Sequenced by Sanger Sequencing
		\item Tags were ``de-convoluted" and counted
	\end{enumerate}
%
%- Conversion of mRNA to cDNA using biotinylated primers
%- Anchored to beads
%- Restriction Enzymes produce 11-mer tags from within a transcript
%- Tags were concatenated before sequencing
%- 99bp contains 9 tags, 550bp contains 50 tags etc

\end{frame}

\begin{frame}{Serial Analysis of Gene Expression}

	\begin{center}
	\includegraphics[scale=0.25]{figures/SAGE.png} 
	\end{center}
	
	\blfootnote{Also see: \url{https://www.scq.ubc.ca/wp-content/uploads/2006/07/SAGE3b.gif}}

\end{frame}

\begin{frame}{Serial Analysis of Gene Expression}

	\begin{itemize}
		\item The word ``tag" is still commonly used in some NGS manuals and software
		\item The term ``Digital Gene Expression" arose during this era
		\begin{itemize}
			\item Is sometimes shortened to DGE, but \textbf{does not} stand for \textit{Differential} Gene Expression.
		\end{itemize}
		\item SAGE doesn't rely on probes targeting known sequences
		\item Variants on the technique are still used\footfullcite{pmid24184689}
		\begin{itemize}
			\item Even used these concatenated tags in early NGS contexts\footfullcite{pmid16721381}
		\end{itemize}
	\end{itemize}

\end{frame}

%\subsection{CAGE}

\begin{frame}{Cap Analysis of Gene Expression}

	\begin{itemize}
		\item A variant technique is \textit{Cap Analysis of Gene Expression}\footfullcite{pmid16489339}
		\item Targets Transcription Start Site (TSS) of mRNA via the 5' cap
		\begin{itemize}
			\item Specifically for identification of the exact TSS and analysis of promoters
		\end{itemize}
		\item Original 27nt long, but now only limited by NGS length
		\item Heavily used in FANTOM (Functional ANnoTation Of the Mammalian genome) project
	\end{itemize}


\end{frame} 

\begin{frame}{SAGE Vs CAGE}

	\begin{itemize}
		\item Primers which target the poly-A sequence will capture \textit{mature} mRNA
		\begin{itemize}
			\item mRNA will also be intact (i.e. not degraded)
		\end{itemize}
		\item CAGE targets transcriptional initiation
		\begin{itemize}
			\item Transcripts may not be ``mature"
			\item 5' Cap must be in place (i.e. not degraded)
		\end{itemize}
		\item \textbf{Both techniques} still involve concatenation of ``tags"
	\end{itemize}

\end{frame}

%\section{Microarray Technology}
%
%\begin{frame}{Microarrays}
%
%	\begin{itemize}
%		\item Microarrays effectively ushered in the modern era of transcriptomics
%		\item Purely interested in \textit{relative abundances}
%		\item Could measure expression levels for 1000’s of genes simultaneously, for \textit{the first time}		
%		\item Were essentially glass slides with probes affixed to them
%	\end{itemize}
%
%\end{frame}
%
\begin{frame}{Microarrays}

	\begin{center}
	\includegraphics[scale=0.5]{figures/timeTrends.png}		
	~\\
	\textcolor[rgb]{0.1,0.4,0.7}{EST (blue)};
	\textcolor[rgb]{0.85,0.65,0.15}{SAGE / CAGE (yellow)};
	\textcolor[rgb]{0.7,0,0.2}{Microarrays (red)};  
	RNA Seq (black) \footfullcite{2017transcriptomictech}
	\end{center}

\end{frame}
%
%\begin{frame}{Microarrays}
%
%	\begin{itemize}
%		\item Once again depends on reverse transcriptase for mRNA $\rightarrow$ cDNA
%		\item \textbf{No reliance on Sanger Sequencing}
%		\item Used probes (like a Northern blot) but the \textbf{cDNA is labelled and the probes are spatially fixed}
%		\begin{itemize}
%			\item Probes must be designed beforehand
%			\item Probes are fixed to the array in \textit{known locations}		
%		\end{itemize}
%	\end{itemize}
%
%\end{frame}
%
%\begin{frame}{Microarrays}
%
%	\begin{enumerate}
%
%		\item Fluorescent labelling during mRNA conversion to cDNA
%		\item Complimentary probes bind target sequences (hybridisation)
%		\item Fluorescence detection at each probe
%	
%	\end{enumerate}
%
%	\begin{center}
%	\textbf{Fluoresence Intensity $\propto$ mRNA abundance	}
%	\end{center}
%
%\end{frame}
%
%\begin{frame}{Microarrays}
%
%	\begin{center}
%	\includegraphics[scale=0.4]{figures/Microarrayhorizontal.png} 
%	\end{center}
%	
%	\blfootnote{Source: \url{https://commons.wikimedia.org/wiki/File:Microarray\_exp\_horizontal.svg}}
%
%\end{frame}
%
%\begin{frame}{Two Colour Microarrays}
%
%	\begin{itemize}
%		\item Probes with known sequences are at known locations
%		\begin{itemize}
%			\item Probes were ~65mer complimentary cDNA
%			\item Originally printed in local facilities
%		\end{itemize}
%		\item Samples are labelled with \textit{either} Cy3 (Green @ 570nm) or Cy5 (Red @ 670nm)
%		\item Both samples are hybridised to array
%		\item Relative Red/Green intensities were of interest
%		\item Gave an estimate of logFC within each array
%	\end{itemize}
%
%\end{frame}
%
%\begin{frame}{Two Colour Microarrays}
%
%	\begin{center}
%	\includegraphics[scale=0.5]{figures/2colour.jpg} 
%	~\\
%	A section of a two colour array\footfullcite{Shalon01071996}
%	\end{center}
%
%\end{frame}
%
%\begin{frame}{Two Colour Microarrays}
%
%	\begin{itemize}
%		\item Probes are ``printed" to the array
%		\begin{itemize}
%			\item Print tips can get clogged
%		\end{itemize}
%		\item Able to be customised for your own experiment
%		\begin{itemize}
%			\item We need a mapping file for probe location to target sequence
%		\end{itemize}
%		\item Both colours were scanned individually
%		\begin{itemize}
%			\item One scan detects red only, the next detects green only
%			\item Each scan would have to be aligned with the other
%		\end{itemize}
%		
%	\end{itemize}
%
%\end{frame}
%
%\begin{frame}{Two Colour Microarrays}
%
%	\begin{itemize}
%		\item Spots were detected using astronomical software
%		\begin{itemize}
%			\item Detection of true signal above background (DABG)
%		\end{itemize}
%		\item ``Spots" could be of variable size
%		\item Dye bias was noted $implies$ experiments often used dye swaps
%		\begin{itemize}
%			\item One sample might be labelled with red on one array, then labelled with green on the next
%		\end{itemize}
%	\end{itemize}
%
%\end{frame}
%
%\begin{frame}{Single Channel Microarrays}
%
%	\begin{itemize}
%		\item 3’ Arrays (Affymetrix) became the dominant transcriptomic technology until RNA seq
%		\item Probes target the 3’ end of transcripts - reduce issues with RNA degradation
%		\item Single channel (i.e. single colour)
%		\item One sample per arrays
%		\item $\sim$1,000,000 $\times$ 25-mer probes
%	\end{itemize}
%
%\end{frame}
%
%\begin{frame}{Single Channel Microarrays}
%
%	\begin{center}
%	\includegraphics[scale=0.25]{figures/Affymetrix.jpg} 
%	\end{center}
%	
%	\blfootnote{Source: \url{https://commons.wikimedia.org/wiki/File:Affymetrix-microarray.jpg}}
%
%\end{frame}
%
%
%\begin{frame}{Single Channel Microarrays}
%
%	\begin{itemize}
%		\item Manufacture used photolithography
%		\item Greater density of probes than two-colour arrays
%		\begin{itemize}
%			\item Shorter probes but far more of them
%		\end{itemize}
%		\item Also need a mapping file from location to probe sequence
%	\end{itemize}
%
%\end{frame}
%
%\begin{frame}{Single Channel Microarrays}
%
%	\begin{center}
%	\includegraphics[scale=0.4]{figures/microarrayLayout.jpg} 
%	\end{center}
%	
%	\blfootnote{Source: \url{https://universe-review.ca/R11-16-DNAsequencing.htm}}
%
%\end{frame}
%
%\begin{frame}{Single Channel Microarrays}
%
%	\begin{itemize}
%		\item Each 3' exon would be targeted by 11 unique probes
%		\item The set of 11 probes would be collected together as a single ``probeset"
%		\item Alternate isoforms with different 3' exons could be detected easily as they would have distinct probesets
%	\end{itemize}
%
%\end{frame}

\end{document}