---
title: "Transcriptomics Applications"
subtitle: "Assignment 2"
output: 
  html_document: 
    toc: yes
    toc_depth: 2
    toc_float: true
---

<style>
div.blue { background-color:#e6f0ff; border-radius: 8px; padding: 12px;}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = FALSE, 
    warning = FALSE,
    message = FALSE
)
```

```{r}
library(tidyverse)
```

# Instructions

## Submission Format [3 marks]

This is assignment is due by 5pm, **Thursday 9^th^ April**.

- Submissions must be made as a *zip archive* containing 2 files:
    1. Your **source R Markdown Document** (with an Rmd suffix)
    2. A **compiled pdf**, showing all code
- All file names **within the zip archive** must start with your student number.
However the name of the zip archive is not important as myUni will likely modify this during submission. [See here for help creating a zip archive](#creating-a-zip-archive)
    
*All questions are to be answered on the same R Markdown / PDF, regardless of if they require a plain text answer, or require execution of code.*

**Marks directly correspond to the amount of time and effort we expect for each question, so please answer with this is mind.**

We **strongly advise** working in the folder `~/transcriptomics/assignment2` on your virtual machine.
Using an R Project for each individual assignment is also strongly advised.

## Creating a zip archive {.tabset}

### On Your VM

If all files required for submission *are contained on your VM*:

1. Select both files using the **Files** pane in R Studio
2. Click export
3. They will automatically be placed into a single zip archive. Please name this in whatever informative name you decide is suitable, but it should contain the suffix `.zip`

### Windows

If all files are on your on your *local Windows machine*:

1. Using File Explorer, enter the folder containing both files
2. Select all files simultaneously by using <kbd>Ctrl</kbd> + Click
3. Right click on one of the files and select `Send to > Compressed (zipped) folder`
4. Rename as appropriate, ensuring the archive ends with the suffix `.zip`

### Mac OS

If all the files are on your *local macOS machine`:

1. Locate the items to zip in the Mac Finder (file system)
2. Right-click on a file, folder, or files you want to zip
3. Select “Compress Items”
4. Find the newly created .zip archive in the same directory and name as appropriate


# Questions

```{r}
marks <- tibble(q = 0, marks = 3)
q <- 1
```


## Question `r q` [3 marks]

```{r}
marks <- marks %>%
    bind_rows(
        tibble(q = q, marks = 3)
    )
```

Many early technologies still have a relevant place in modern transcriptomics.
Of the technologies covered in *Lecture 2: Early Transcriptomic Strategies*, which technology might be suitable for analysis of pri-miRNAs?
Explain why in one or two brief sentences.

<div class = "blue">

- CAGE is actively used in study of miRNA as it is able to distinguish the multiple pri-miRNAs which lead to identical miRNA
- This takes advantage of the 5' cap and can uniquely identify the locus of origin

</div>

```{r}
q <- q + 1
```

## Question `r q` [8 marks]

```{r}
marks <- marks %>%
    bind_rows(
        tibble(q = q, marks = 8)
    )
```

Briefly contrast **two** of the technologies presented in *Lecture 2: Early Transcriptomic Strategies* and *Lecture 3: Microarrays*.
Discuss their strengths and limitations, paying particular attention to how each represented a breakthrough at the time they gained prominence.

<div class = "blue">

Some possible discussion points are

- Northern blot:
    - Limited sequences are able to be probed for
    - Relative Quantiation is crude
    - Alternate isoforms can be identified
- ESTs:
    - The sequences of expressed transcript were able to be determined for the first time using a high-throughput approach
    - No accurate quantification was able to be performed and throughput was still limited to quantities which were feasible for Sanger-sequencing 
- CAGE/SAGE:
    - Process is very complicated
    - Relatively accurate quantitation (compared to microarrays)
    - Sequences do not need to be known *a priori*
- RT-qPCR:
    - Used for single genes or targeted sequences $\implies$ low throughput
    - Accurate quantitation
- Microarrays:
    - Expression estimates and the relative dynamics were now able to be quantified in the thousands
    - Genes were limited to those identified at design time
    - The dynamic range was restricted by the limits of the scanner
    - LImited ability to idenfity multiple isoforms
    - Background noise was problematic

</div>

```{r}
q <- q + 1
```

## Question `r q` [3 marks]

```{r}
marks <- marks %>%
    bind_rows(
        tibble(q = q, marks = 3)
    )
```

When performing a statistical test, we usually frame our analysis in terms of a *Null Hypothesis* and an *Alternate Hypothesis*, eventually returning a $p$-value.
Explain why we do this, including clear description of what a $p$-value represents?

<div class = "blue">

- The Null Hypothesis ($H_0$) gives us a chance to describe what the data should like like if **nothing is happening**. By implication anything that is observed, but that appears unlikely under $H_0$ means that **something is happening**.
- A p-value represents the probability of observing data as (or more) extreme than we have observed if the Null Hypothesis were true

</div>

```{r}
q <- q + 1
```

## Question `r q` [3 marks]

```{r}
marks <- marks %>%
    bind_rows(
        tibble(q = q, marks = 3)
    )
```

When conducting a $T$-test, we estimate two population-level parameters.
Describe both of these, using the context of comparing gene expression patterns across two treatment groups.

<div class = "blue">

- The population mean $\mu$ and the population variance $\sigma^2$
- In the context of gene expression:
    - $\mu$ is the average expression level across all possible individuals
    - $\sigma$ is the variability in expression level across all possible individuals

</div>

```{r}
q <- q + 1
```

## Question `r q` [10 marks]

```{r}
marks <- marks %>%
    bind_rows(
        tibble(q = q, marks = 10)
    )
```

To obtain your own set of qPCR data, please execute the following lines of code, **using your own student number instead of the example given** (`"a1234567"`).
This will create an object called `qPCR` in your R Environment.
This object will contain $C_t$ values for a *gene of interest* and a *housekeeper* gene, across two cell types.
Each experiment is run as a series of pairs so that you will have 4 values for each pair (2X Cell Types + 2X Genes).

```{r, echo=TRUE, eval=TRUE}
source("https://uofabioinformaticshub.github.io/transcriptomics_applications/assignments/A2Funs.R")
makeRT("a1234567")
```


For this question, please perform the following tasks, showing all code.
Where suitable, use `pander()` to display the results.

1. Generate a clearly labelled boxplot for each gene and cell type. 
2. After generation of the boxplot, comment on the suitability of the housekeeper, paying particular note to the stability of it's expression between your cell types.
3. Calculate the $\Delta C_t$ values for your gene of interest by normalising to the housekeeper
4. Calculate the $\Delta \Delta C_t$ values for the comparison between cell types
5. Using the above $\Delta \Delta C_t$ values, perform a $T$-test using the R function `t.test()` and interpret the output

<div class = "blue">

```{r, echo=TRUE, fig.align='center'}
qPCR
qPCR %>%
  ggplot(aes(gene, Ct, fill = gene)) +
  geom_boxplot() +
  facet_wrap(~cellType) +
  theme_bw()
```

- For this dataset, the expression of the housekeeper *RPL19* appears relatively stable across both cell types. This may not have been the case for some students' data

Calculation of dCT using `pivot_wider()` to rearrange columns. 
There are numerous alternative approaches, but this was the one I chose.

```{r, echo=TRUE}
dCT <- qPCR %>%
  pivot_wider(
    id_cols = c(cellType, pair),
    names_from = gene,
    values_from = Ct
  ) %>%
  mutate(
    dCt = FOXO1 - RPL19
  )
dCT
```

Calculation of ddCT using a similar strategy

```{r, echo=TRUE}
ddCT <- dCT %>%
  pivot_wider(
    id_cols = pair,
    names_from = cellType,
    names_prefix = "dCt_",
    values_from = dCt
  ) %>%
  mutate(
    ddCT = dCt_Th - dCt_Treg
  )
ddCT
```

To perform the t-test

```{r, echo=TRUE}
t.test(ddCT$ddCT)
```

For $H_0:$ *true average change in FOXO1 expression is zero between Th and Treg*, the t-test produced a p-value of 0.2014.
As this is $>0.05$ we would accept $H_0$ and conclude that the true average difference in expression is zero.


</div>


```{r}
q <- q + 1
```

## Question `r q` [10 marks]

```{r}
marks <- marks %>%
    bind_rows(
        tibble(q = q, marks = 10)
    )
```

For this question, you will need the objects `cpm.tsv`, `topTable.csv` and `de.tsv`.
You will be assigned a *gene set* using the following command, again remembering to use your own student ID number.

```{r, echo=TRUE, eval=FALSE, results='hide'}
source("https://uofabioinformaticshub.github.io/transcriptomics_applications/assignments/A2Funs.R")
chooseGeneSet("a1234567")
```

For this question, your task is to:

1. Find which *differentially expressed* genes belong to this gene-set.
These are provided in the object `de.tsv`, and these should be **formed into a `character` vector**.
2. Restrict this character vector so that it only contains genes within `cpm.tsv`. (Hint: You can use `topTable.csv` to map from gene names to gene IDs)
3. Using `pheatmap()`, create a heatmap of these genes using the expression values contained in `cpm.tsv`.
For reference, these values are provided as `logCPM` values, which are suitable for plotting directly.
Include an annotation for each sample, indicating which genotype it represents. 

<div class = "blue">

```{r, echo=TRUE}
chooseGeneSet("a1234567")
```

First we need to find which DE genes belong to this geneset

```{r, echo=TRUE}
de <- here::here("practicals/data/de.tsv") %>%
  read_tsv() %>%
  dplyr::filter(gs_name == "GO_RESPONSE_TO_CYTOKINE") 
deGenes <- str_split(de$DE, pattern = ";")[[1]]
deGenes
```

Given that the object `cpm.tsv` and `topTable.csv` contain only a subset of genes, we can subset our vector of DE genes.

```{r, echo=TRUE}
topTable <-  here::here("practicals/data/topTable.csv") %>%
  read_csv()
myDeGenes <- deGenes[deGenes %in% topTable$gene_name]
myDeGenes
```

Now we can use these to create a heatmap

```{r, echo=TRUE, fig.align='center'}
logCPM <-  here::here("practicals/data/cpm.tsv") %>%
  read_tsv()
```

```{r, echo=TRUE}
library(pheatmap)
library(scales)
topTable %>%
  dplyr::filter(gene_name %in% myDeGenes) %>%
  dplyr::select(gene_id, gene_name) %>%
  left_join(logCPM) %>%
  dplyr::select(gene_name, starts_with("Ps")) %>%
  rename_all(
    str_replace, 
    pattern = ".+(Het|WT).+_F3_([0-9]+).+", 
    replacement = "\\1_\\2"
  ) %>%
  as.data.frame() %>%
  column_to_rownames("gene_name") %>%
  pheatmap(
    color = viridis_pal(option = "magma")(100),
    cutree_cols = 2,
    cutree_rows = 2,
    angle_col = 0
  )
```




</div>


```{r}
q <- q + 1
```


# Total: `r sum(marks$marks)` marks


```{r results, results='as.is'}
late <- c("a17455555", "a1724305")
here::here("assignments", "A2.csv") %>%
  read_csv() %>%
  rename(points = marks) %>%
  pivot_longer(
    cols = starts_with("a"),
    names_to = "Student ID",
    values_to = "awarded"
  ) %>%
  dplyr::filter(!is.na(awarded)) %>%
  group_by(`Student ID`) %>%
  summarise(
    Marks = sum(awarded),
    `Scaled (10)` = 10 * sum(awarded) /sum(points)
  ) %>%
  left_join(
    here::here("practicals/VMs.csv") %>%
      read_csv
  ) %>%
  dplyr::select(
    contains("Student"), Marks, `Scaled (10)`
  ) %>%
  mutate(
    Penalty = case_when(
      `Student ID` %in% late ~ 0.1*Marks,
      !`Student ID` %in% late ~ 0
    ),
    Total = Marks - Penalty,
    `Scaled (10)` = case_when(
      `Student ID` %in% late ~ (1 - 0.1)*`Scaled (10)`,
      !`Student ID` %in% late ~ `Scaled (10)`
    ),
    Total = round(2*Total, 0) / 2,
    `Scaled (10)` = round(`Scaled (10)`, 2),
    Grade = cut(
      `Scaled (10)`, 
      breaks = c(0, 5, 6.5, 7.5, 8.5, 10), 
      labels = c("F", "P", "C", "D", "HD")
      )
  ) %T>%
  write_csv(here::here("assignments/A2_results.csv")) %>%
  arrange(desc(Total)) %>%
  # pander::pander(split.table = Inf)# %>%
  summarise(
    highest = max(`Scaled (10)`),
    median = median(`Scaled (10)`)
  ) %>%
  pander::pander(caption = "Summary of grades for assessment 2 2020, out of a possible 10 marks")
```
